lib/jxl/splines.cc

```diff
@@ -567,13 +571,15 @@ Status Splines::InitializeDrawCache(const size_t image_xsize,
   size_t px_limit = (pixel_limit < static_cast<float>(kHardPixelLimit))
                         ? static_cast<size_t>(pixel_limit)
                         : kHardPixelLimit;
+  std::vector<Spline::Point> intermediate_points;
   for (size_t i = 0; i < splines_.size(); ++i) {
     JXL_RETURN_IF_ERROR(
         splines_[i].Dequantize(starting_points_[i], quantization_adjustment_,
                                cmap.YtoXRatio(0), cmap.YtoBRatio(0), spline));
     if (std::adjacent_find(spline.control_points.begin(),
                            spline.control_points.end()) !=
         spline.control_points.end()) {
+      // TODO(eustas): spec does not explicitly say this is a problem...
```

Comment:
It does - it states that whenever a division by 0 would be required, the number you divide by is non0 in a well-formed codestream.

Couldn't we instead check that the difference between consecutive points is "large enough"?
@sboukortt for opinions

Reply:
I mean, it says that we should render Cutmull-Rom spline, and gives some example code. However, mathematically, division by zero is not required if we carefully interprete what happens in the situation of repeated control points.
I see no reason to forbid cusps (spinodes) - those could be useful, e.g. for text rendering.

Reply:
IIRC the "example code" is not an example, but it is supposed to be exactly what needs to happen...

Reply:
I'm fine with allowing repeated control points as long as nothing changes in the rendering of cases without repeated control points. We can update the spec in that direction as far as I'm concerned. But it's indeed not just "example code", it is the specification of how the decoded image should look like, and any other implementation needs to produce the same pixels.

Reply:
On the second thought - we allow control points to be fairly close to each other, so that visual cusps are achievable. So, let's leave spec intact and just deal with FP precision problems.

Reply:
We use the centripetal variant of Catmull-Rom splines so cusps should not happen: http://www.cemyuksel.com/research/catmullrom_param/ (α=0.5, blue curves)

Reply:
Ah, I see what's going on and this change is indeed a noop with infinite precision, so that seems fine - except for the case of a single point...